test/function_reference.js

```diff
@@ -1,20 +1,157 @@
+            events.push({ eventName: 'init', type, triggerAsyncId, resource });
+          }
+        },
+        before (asyncId) {
+          if (asyncId === id) {
+            events.push({ eventName: 'before' });
+          }
+        },
+        after (asyncId) {
+          if (asyncId === id) {
+            events.push({ eventName: 'after' });
+          }
+        },
+        destroy (asyncId) {
+          if (asyncId === id) {
+            events.push({ eventName: 'destroy' });
+            destroyed = true;
+          }
+        }
+      })
+      .enable();
+  });
+}
+
+function canConstructRefFromExistingRef (binding) {
+  const testFunc = () => 240;
+  assert(binding.ConstructWithMove(testFunc) === 240);
+}
+
+function canCallFunctionWithDifferentOverloads (binding) {
+  let outsideRef = {};
+  const testFunc = (a, b) => a * a - b * b;
+  const testFuncB = (a, b, c) => a + b - c * c;
+  const testFuncC = (a, b, c) => {
+    outsideRef.a = a;
+    outsideRef.b = b;
+    outsideRef.c = c;
+  };
+  const testFuncD = (a, b, c, d) => {
+    outsideRef.result = a + b * c - d;
+    return outsideRef.result;
+  };
+
+  assert(binding.CallWithVec(testFunc, 5, 4) === testFunc(5, 4));
+  assert(binding.CallWithInitList(testFuncB, 2, 4, 5) === testFuncB(2, 4, 5));
+
+  binding.CallWithRecvVector(testFuncC, outsideRef, 1, 2, 4);
+  assert(outsideRef.a === 1 && outsideRef.b === 2 && outsideRef.c === 4);
+
+  outsideRef = {};
+  binding.CallWithRecvInitList(testFuncC, outsideRef, 1, 2, 4);
+  assert(outsideRef.a === 1 && outsideRef.b === 2 && outsideRef.c === 4);
+
+  outsideRef = {};
+  binding.CallWithRecvArgc(testFuncD, outsideRef, 2, 4, 5, 6);
+  assert(outsideRef.result === testFuncD(2, 4, 5, 6));
+}
+
+async function canCallAsyncFunctionWithDifferentOverloads (binding) {
+  const testFunc = () => 2100;
+  const testFuncB = (a, b, c, d) => a + b + c + d;
+  let hook = installAsyncHook();
+  binding.AsyncCallWithInitList(testFunc);
+  let triggerAsyncId = asyncHook.executionAsyncId();
+  let res = await hook;
+  assert.deepStrictEqual(res, [
+    {
+      eventName: 'init',
+      type: 'func_ref_resources',
+      triggerAsyncId: triggerAsyncId,
+      resource: {}
+    },
+    { eventName: 'before' },
+    { eventName: 'after' },
+    { eventName: 'destroy' }
+  ]);
+
+  hook = installAsyncHook();
+  triggerAsyncId = asyncHook.executionAsyncId();
+  assert(
+    binding.AsyncCallWithVector(testFuncB, 2, 4, 5, 6) === testFuncB(2, 4, 5, 6)
+  );
+  res = await hook;
+  assert.deepStrictEqual(res, [
+    {
+      eventName: 'init',
+      type: 'func_ref_resources',
+      triggerAsyncId: triggerAsyncId,
+      resource: {}
+    },
+    { eventName: 'before' },
+    { eventName: 'after' },
+    { eventName: 'destroy' }
+  ]);
+
+  hook = installAsyncHook();
+  triggerAsyncId = asyncHook.executionAsyncId();
+  assert(
+    binding.AsyncCallWithArgv(testFuncB, 2, 4, 5, 6) === testFuncB(2, 4, 5, 6)
+  );
+}
+function test (binding) {
```

Comment:
This should be async because it should await  canCallAsyncFunctionWithDifferentOverloads.

Reply:
... and it should itself be `await`ed at the call site.